py-rse/correct.Rmd

```diff
@@ -528,45 +523,98 @@ bin/test_zipfs.py ..                                                     [100%]
-and the second involves estimating the \(\alpha\) parameter from the word count.
-Our unit tests `test_word_count` and `test_alpha` have checked that these two
-individual components work in isolation, but do they work correctly together?
-Checking that multiple units work correctly together is called *integration testing*.
+2.  If `func` raises the error it's supposed to
+    then we go into the first `except` branch
+    *without* triggering the `assert` immedately below the function call.
+    The code in this `except` branch could check that
+    the exception contains the right error message,
+    but in this case it does nothing
+    (which in Python is written `pass`).
 
-The structure of integration tests is very similar to that of unit tests.
-There is an expected result, which is compared against the actual result.
-However, the work involved in creating the expected result
-and/or setting up the code to run
-can be considerably more complicated and involved. 
+3.  Finally,
+    if the function raises the wrong kind of exception
+    we also `assert False`.
+    Checking this case might seem overly cautious,
+    but if the function raises the wrong kind of exception,
+    users could easily fail to catch it.
+
+This pattern is so common that `pytest` provides support for it.
+Instead of the eight lines in our original example,
+we can instead write:
+
+```python
+import pytest
+
+...set up fixture...
+with pytest.raises(ExpectedError):
+    actual = func(fixture)
+```
+
+The argument to `pytest.raises` is the type of exception we expect.
+The call to the function then goes in the body of the `with` statement.
+We will explore `pytest.raises` further in the exercises.
+
+## Integration Testing {#correct-integration}
+
+Our Zipf's Law analysis has two steps:
+counting the words in a text
+and estimating the \(\alpha\) parameter from the word count.
+Our unit tests give us some confidence that these components work in isolation,
+but do they work correctly together?
+Checking that is called [integration testing][integration-testing].
+
+Integration tests are structured the same way as unit tests:
+a fixture is used to produce an actual result
+that is compared against the expected result.
+However,
+creating the fixture and running the code
+can be considerably more complicated. 
 For example,
 in the case of our Zipf's Law software an appropriate integration test fixture
 might be a text file with a word frequency distribution that has a known alpha value. 
 In order to create this text fixture,
-we'll need a way of generating random words.
-Fortunately, a Python library exits to do just that.
-We can install it (and the `pypandoc` library it depends on) at the command line
+we need a way to generate random words.
+
+Fortunately, a Python library called `randomwordgenerator` exists to do just that.
+We can install it and the `pypandoc` library it depends on
 using the Python Package Installer (pip): 
```

Comment:
Should "Python Package Installer" have a glossary entry?

Reply:
:+1: added